### ISSUE TITLE
Self module reference

### ISSUE THREAD

**Issue by alexi2014** [rollup/rollup#1055]
Importing of a module from itself like here:
**file module.js**
`import * as m from './module';`

causes "A module cannot import itself" error. 
According to this link this should work better than ever:
[https://github.com/rollup/rollup/wiki/Cycles](https://github.com/rollup/rollup/wiki/Cycles)

Seems to be a bug?

**Comment by Rich-Harris** [comment]
Cycles are when module A imports module B, and module B imports module A. If you're in `module.js` what reason could you have for importing things that are defined in that module? They're already there! (Genuine question, I'm curious as to whether there's a use case here.)

I'm not sure if self-imports are forbidden by the spec, but I can't imagine a situation where this wouldn't be a bug that should cause Rollup to stop work and inform the developer.

**Comment by alexi2014** [comment]
Regarding a use case, we implement a "soft" module dependency using it, something like
if (moduleA is loaded) {
     use moduleA
}
As you know, this is impossible using ES6 modules import/export syntax, and System object that could allow it is not approved yet, even for ES7. And our implementation passes through all well known bundlers that tends to collect all possible "import" calls, or report an error if such a call is too tangled for them to resolve (and there are objective reasons for this).
If you're interested in more details, I can expand. Briefly, a module gets a reference to itself, registers this reference using our private means, and another module can check (using private api as well) whether another one is loaded, and get a reference to it if it's the case.

Regarding my confidence in that a self-reference should work, I have two points:
1) All major bundlers/loaders support this, both rather old and the newest ones: Browserify, RequireJS, WebPack, SystemJS.
2) I didn't find any word about that it's prohibited in ES6 spec.

Thanks,
Alex

**Comment by Rich-Harris** [comment]
Could you illustrate what you're saying with some code (and the bundle you'd expect that code to generate, or the code that other bundlers generate) please? Afraid I'm finding it very hard to follow!

**Comment by alexi2014** [comment]
Here are the samples that demonstrate self-references support in the aforementioned bundlers/loaders:
[https://www.dropbox.com/s/t8bwz6hzyel1b2r/ModuleSelfReference.zip?dl=0](https://www.dropbox.com/s/t8bwz6hzyel1b2r/ModuleSelfReference.zip?dl=0)

The samples are:
Bundlers:
**CommonJsModuleBrowserifyTest** - Browserify with CommonJS modules
**AmdModuleWebpackTest** - WebPack with AMD modules
**CommonJsModuleWebpackTest** - WebPack with CommonJS modules

Loaders:
**AmdModuleTest** - RequireJS with AMD modules
**SystemModuleJspmTest** - Jspm (loader) with System modules

Each sample contains two modules. **module1.js** contains a self-reference, and assigns it to a global variable. **module2.js** imports module1, and gets references to module1 in two ways - via a standard reference, and via the global variable assigned by the module1. And shows a content obtained from the module1 via these two kinds of references on a page.
This is not exactly what we do in our library, but it demonstrates the basics of this functionality.

Each sample demonstrating a bundler already contains a generated bundle. If you want to create bundles by yourself, please follow the instructions in the accompanying readme.txt files.
To run a sample, just open index.html in a browser. For samples demonstrating loaders you'll need to host this page in a web server.

Alex

**Comment by rastographics** [comment]
I think this "bug" may be causing the [issue I have here](https://github.com/Breeze/breeze.js/issues/170)

Don't know why it's telling me that breeze is importing itself. I can't find the self-reference anywhere in the breeze source code.

**Comment by alexi2014** [comment]
Sorry for boring you, is there any update on this? Does it make sense to wait for an answer?

Thanks,
Alex

**Comment by rastographics** [comment]
@alexi2014 in my case I ended up working around this issue by removing the self-reference in order to get rollup to build.

**Comment by rafabernad** [comment]
Hi,

Just wanted to comment that self imports should be allowed, talking about ES6 compliance:

http://stackoverflow.com/questions/40240659/how-can-an-es6-module-import-itself

**Comment by jcmoore0** [comment]
It would be great.  I would like to  try using rollup with Wijmo. 
Best regards,
John

**Comment by olsonpm** [comment]
To address the use-case Rich, I asked Bergi just that on stack overflow the other day:

https://stackoverflow.com/questions/40240659/how-can-an-es6-module-import-itself/40242291#comment75093422_40242291

His answer was less about semantics, but more about the useful functionality it provides.

**Comment by nickaranz** [comment]
Any updates on supporting self-reference?

**Comment by cmcaine** [comment]
This is also preventing me from using rollup :(

I had a quick look at the rollup code, but it looks like a lot of work for someone new to the codebase to fix this, but maybe not much work for someone familiar with it?

My use case is that I want to retrieve attributes of my module scope dynamically to avoid writing boilerplate:

```js
// example.js
export function foo() {}
export function bar() {}

import * as SELF from './example'
receiveInterProcessMessage(message) {
    SELF[message.command](...message.args)
}
```

Here are some workarounds:

```js
// Option 1
export function foo() {}
export function bar() {}

// But then I have to maintain this list separately from the functions
messagable = {
	foo,
	bar,
}

// Option 2
const messagable = {}
function ipc(func) {
    messagable[func.name] = func
}

// But this is kind of verbose
export const foo = ipc(function foo() {})
export const bar = ipc(function bar() {})

// Option 3, with a sweet.js macro that will populate messagable (but sweet.js doesn't support TS yet).
messagable = {}
ipcfunc foo() {}
ipcfunc bar() {}
```

Anyway my desired transliteration of the self import in my first example:

```js
SELF = Object.freeze(Object.create(null, {
    // List of exports of this module
    foo: foo1 // rollup's generated names here
    bar: bar3
    // ...
}))
```

And rollup should do the same static analysis it does to other module imports to ensure that properties of SELF are never written to.

non-star imports of your own module are pretty pointless, but could be supported through re-writing the source file before other transformations:

```js
// Foo is a const, so this should fail
import {foo} from './example'

import {foo as FOO} from './example'
// Rewritten to...
const FOO = foo

export {foo as FOO2} from './example'
// Rewritten to...
export const FOO2 = foo
```

**Comment by guybedford** [comment]
This is a perfectly valid import case and will work in all implementations and the spec. Also just hitting this now.

**Comment by lukastaegert** [comment]
I agree that this limitation is somewhat artificial. For now, the main issue I see is to identify non-hoisted (i.e. `let` or `const`) declarations that are part of a cyclic or self reference and rewrite them with `var`. Beyond that I think the algorithm should actually be able to handle self references already. Maybe we can do this as part of a broader move to fix cyclic dependency issues. The latter, however, will also require improvements to the module sorting algorithm.

**Comment by guybedford** [comment]
@lukastaegert thanks for the update here.

I'm not sure I follow why a replacement is necessary as in the following code:

self.js
```js
import { x } from './self.js';
console.log(x);
export const x = 5;
```

it would be correct for it to get a thrown error on unable to access lexical declaration `x` as `x` has not been defined yet.

So the straightforward Rollup binding transformation of the above should be fine I believe?

**Comment by lukastaegert** [comment]
I guess we should take another look at the spec. Can someone confirm how this is handled by e.g. latest node versions or browsers supporting ES modules? Do they throw an error or do they print undefined? I agree that the latter would be a little weird considering it is a constant.

**Comment by guybedford** [comment]
@lukastaegert I can confirm the above behaviour is how browsers work and is by the spec. An explicit error is thrown when accessing a const or let binding before its initialization statement has been executed. For more info, read up on TDZ (temporal dead zone).

**Comment by lukastaegert** [comment]
Of course we should change the example to use different names so that it does not fail due to an internal TDZ problem:

```javascript
import { x as y } from './self.js';
console.log(y);
export const x = 5;
```

However, I would expect this to work:
```javascript
import { x as y } from './self.js';
Promise.resolve().then(() => console.log(y));
export const x = 5;
```

Or am I wrong here?

**Comment by lukastaegert** [comment]
The latter would not survive scope-hoisting if we do not either change `const` to `var` or introduce another variable in between.

**Comment by guybedford** [comment]
@lukastaegert in the first case, as the module system is specified, `y` would still be the same _binding_ as `x` so would still throw the TDZ error identically. Similarly, `export default 5` is treated as a let-style binding so would also throw on early access for these sorts of cases.

In the second case, yes the execution of `const x = 5` has happened by the time the promise resolves, so there is no issue.

**Comment by guybedford** [comment]
(the second case being the same as `Promise.resolve().then(() => console.log(x)); const x = 5` which would also work)

**Comment by lukastaegert** [comment]
Ah I see my mistake here. So it seems it should probably work already if we just remove the hard coded check that is throwing the "cannot import self" error. Still I wonder why this check was introduced in the first place, maybe it was more or less to serve as documentation for otherwise harder to understand errors?

 I am a little short on time these days but I hope to find some time soon to look through old commits to see at least when this check was added. Otherwise, it might be actually  quite easy to fix this.

**Comment by guybedford** [comment]
Great to hear that, yes exactly the way Rollup works in _inlining bindings_ works beautifully and naturally with all these cases, they should all work out without much effort in Rollup apart from ensuring correct binding rereferencing and execution ordering. Perhaps @Rich-Harris wasn't aware that these cases matched up with the spec exactly this way!

I'd advocate following the spec here, it's a luxury this project has that many others don't, and it's function is to resolve disputes such as these!

**Comment by guybedford** [comment]
@lukastaegert will do a PR if it helps.
